tests/force_insert_update/tests.py

```diff
@@ -66,3 +77,153 @@ def test_force_update_on_inherited_model_without_fields(self):
+            Parent(id=5, name="Child-5", value=12).save(force_insert=True)
+
+        GrandParent(id=6, name="Grandparent-6", value=12).save()
+        with self.assertNumQueries(3):
+            Parent(id=6, name="Child-6", value=12).save()
+
+    def test_force_insert_on_inherited_model_grandparent_exists_parent_exists(self):
+        with transaction.atomic():
+            GrandParent(id=1, name="Grandparent-1", value=8).save()
+            Parent(id=1, name="Parent-1", value=8).save()
+            with self.assertRaises(IntegrityError):
+                Child(id=1, name="Child-1", value=8).save(force_insert=models.Model)
+
+        with transaction.atomic():
+            GrandParent(id=2, name="Grandparent-2", value=9).save()
+            Parent(id=2, name="Parent-2", value=9).save()
+            with self.assertRaises(IntegrityError):
+                Child(id=2, name="Child-2", value=9).save(
+                    force_insert=(Parent, GrandParent)
+                )
+
+        with transaction.atomic():
+            GrandParent(id=3, name="Grandparent-3", value=10).save()
+            Parent(id=3, name="Parent-3", value=10).save()
+            with self.assertRaises(IntegrityError):
+                Child(id=3, name="Child-3", value=10).save(force_insert=(Parent,))
+
+        # Backward compatibility tests
+        GrandParent(id=4, name="Grandparent-4", value=11).save()
+        Parent(id=4, name="parent-4", value=11).save()
+        with self.assertNumQueries(3):
+            Child(id=4, name="Child-4", value=11).save(force_insert=True)
+
+        GrandParent(id=5, name="Grandparent-5", value=11).save()
+        Parent(id=5, name="parent-5", value=11).save()
+        with self.assertNumQueries(4):
+            Child(id=5, name="Child-5", value=11).save()
+
+    def test_create_query_count(self):
+        GrandParent.objects.create(id=1, name="Grandparent-1", value=8)
+        with self.assertNumQueries(2):
+            Parent.objects.create(id=1, name="Parent-1", value=8)
+
+        with self.assertNumQueries(3):
+            Parent.objects.create(id=2, name="Parent-2", value=8)
+
+    def test_force_insert_in_diamond_inheritance(self):
+        # p, gp DNE
+        with transaction.atomic():
+            with self.assertRaises(IntegrityError):
+                DiamondChild(id=2, name="Child-2", value=8).save(
+                    force_insert=(GrandParent, Parent, DiamondParent)
+                )
+
+        with transaction.atomic():
+            with self.assertRaises(IntegrityError):
+                DiamondChild(id=3, name="Child-3", value=8).save(
+                    force_insert=(models.Model,)
+                )
+
+        # gp E, p DNE
+        GrandParent(id=6, name="Grandparent-6", value=8).save()
+        with transaction.atomic():
+            with self.assertRaises(IntegrityError):
+                DiamondChild(id=6, name="Child-6", value=8).save(
+                    force_insert=(models.Model,)
+                )
```

Comment:
Are these tests failing due to #16903 and #16952? If it's the case then I think it's a strong argument to merge the other two first so we can have proper testing in place here.

Reply:
This is expected scenario, as the `GrandParent` already exists and we are trying to force insert that again it fails. This is the reason the force insert flag was not passed to `_save_parents` in the first place.

But it completely makes sense to pass #16952 and #16903 first before this one, as they are major bugs and keeps almost half the features of Django MTI unavailable.

In case of #16952 I am in completely blank as I am getting a completely unexpected error there. Your insights on this [comment](https://github.com/django/django/pull/16952#issuecomment-1581806139) willbe very much helpful.